### PR TITLE
feat(forms): add form related methods

### PR DESCRIPTION
**Pull request by basvandenheuvel**
@robwormald

Credits to @MarkPerryBV for helping out with this PR.
#### Changes in this PR

AbstractControl: 
- Added `markAsUntouched()`, `markAsPristine()`

ControlGroup: 
- Added `markAsPristine()`, `markAllAsPristine()`, `markAllAsDirty()`, `markAllAsTouched()`, `markAllAsUntouched()`, `markAllAsPending()`

ControlArray: 
- Added `markAsPristine()`, `markAllAsPristine()`, `markAllAsDirty()`, `markAllAsTouched()`, `markAllAsUntouched()`, `markAllAsPending()`

Tests:
- Added missing tests for "touched" part
- Added tests for all new methods

NG_COMMON:
- Added needed NG_COMMON values (ask permission? @mhevery)
#### Questions for the Angular team:
- Methods like "dirty" and "pristine" bubble up to the parent, the touched part doesn't. Is this intended? If the touched part should also bubble up i can create a PR for this, let me know.
#### Next PR
- For a next PR i'm going to look into stuff like `clear()` and `reset()` methods for the form.

Issues I (partly) solve with this PR: #4933, #5568

### CODE REVIEW

**Comment by escardin** [comment]
@basvandenheuvel I'm not sure I understand the point of the non-all methods. Wouldn't a ControlGroup/ControlArray only be pristine,dirty,touched, etc... if at least one control within it is that way? This would imply that markAsPristine() should make all the controls within it pristine.

I was also under the impression that ControlGroup and ControlArray also inherit/implement AbstractControl, and as such, you've extended their api in a way that users may find unexpected. Consider the situation with a ControlGroup nested inside another ControlGroup or ControlArray (or vice versa). You have handled this situation by extra code which checks if it's an array or group, when you could have just called the common method.

**Comment by basvandenheuvel** [comment]
@escardin 

*"Wouldn't a ControlGroup/ControlArray only be pristine,dirty,touched, etc... if at least one control within it is that way? "*
- That does apply to dirty. If one child is dirty, the Group / Array is dirty.
- That does not apply to touched. See the source code, the touched part does not bubble up.
*( I don't know why, and as you can see in my PR description i asked this to the NG2 team.)*
- That DOES NOT apply to pristine. If one of the children is pristine, but all others are dirty, the Group / Array is still dirty.

*"This would imply that markAsPristine() should make all the controls within it pristine."*
- There never was a way to make stuff go down the tree. This is something i introduced using the `markAllAs*()` methods. Check the source code, when you call `markAsDirty()` on a group (their code, not mine) It only marks the Group as dirty and the parent, it doesn't go down the tree.

*"Consider the situation with a ControlGroup nested inside another ControlGroup or ControlArray (or vice versa). You have handled this situation by extra code which checks if it's an array or group, when you could have just called the common method."*
- One of the reason i introduced the `markAllAs*()` methods is because you might have nested Groups. If a Group calls the default `markAsDirty()` method on his child Group that child Group is set to dirty and that method tries to make the parent dirty aswell, it doesn't go down the tree....

- Check my source code to see what happens when you call `markAllAsDirty()` on a Group containing Groups. Very simple, this is what it does:
  - 1. Loop through all childeren
    - 1.1 Is it a Control then call `markAsDirty({onlySelf: true})`
    - 1.2 Is it a ControlGroup or ControlArray then call `markAllAsDirty({onlySelf: true})` and for that Group / Array we start at step 1.
  - 2. Call the super `markAsDirty({onlySelf: onlySelf})` on the Group/Array.

**Comment by basvandenheuvel** [comment]
@escardin so your last sentence "could have called the common method" is not possible because there was no way of going down the tree. And implementing that stuff in AbstractControl was not possible either since Control also inherits that class..

**Comment by robwormald** [comment]
cc @vsavkin

**Comment by iorlas** [comment]
Wow, looks cool. Guys, we really need this. Thank you, I'm eager to hop on angular 2 train, but I can't without this PR.

**Comment by MarkPerryBV** [comment]
Any update on the code review guys?

**Comment by vsavkin** [comment]
Could you squash all the commits?

File: modules/angular2/src/common/forms/model.ts [hunk 1]
```diff
@@ -367,6 +378,101 @@ export class ControlGroup extends AbstractControl {
...
+   * Mark the group and all it's children as pristine.
+   */
+  markAllAsPristine({onlySelf}: {onlySelf?: boolean} = {}): void {
+    onlySelf = normalizeBool(onlySelf);
+
+    StringMapWrapper.forEach(this.controls, (control, name) => {
+      if (control instanceof Control) {
+        control.markAsPristine({onlySelf: true});
+      } else if (control instanceof ControlGroup || control instanceof ControlArray) {
+        control.markAllAsPristine({onlySelf: true});
```
Comment (vsavkin) [inline modules/angular2/src/common/forms/model.ts]: Could you either add an else clause that throws or rewrite the if statement as follows:

```
if (control instanceof ControlGroup || control instanceof ControlArray) {
  control.markAllAsPristine({onlySelf: true});
} else {
  control.markAsPristine({onlySelf: true});
}
```

File: modules/angular2/src/common/forms/model.ts [hunk 2]
```diff
@@ -367,6 +378,101 @@ export class ControlGroup extends AbstractControl {
...
+  }
+
+  /**
+   * Mark the group and all it's children as dirty.
+   */
+  markAllAsDirty({onlySelf}: {onlySelf?: boolean} = {}): void {
+    onlySelf = normalizeBool(onlySelf);
+
+    StringMapWrapper.forEach(this.controls, (control, name) => {
+      if (control instanceof Control) {
```
Comment (vsavkin) [inline modules/angular2/src/common/forms/model.ts]: Same comment about the if statement.

File: modules/angular2/src/common/forms/model.ts [hunk 3]
```diff
@@ -367,6 +378,101 @@ export class ControlGroup extends AbstractControl {
...
+    });
+
+    super.markAsDirty({onlySelf: onlySelf});
+  }
+
+  /**
+   * Mark the group and all it's children as touched.
+   */
+  markAllAsTouched(): void {
+    StringMapWrapper.forEach(this.controls, (control, name) => {
```
Comment (vsavkin) [inline modules/angular2/src/common/forms/model.ts]: Same comment about the if statement.

File: modules/angular2/src/common/forms/model.ts [hunk 4]
```diff
@@ -367,6 +378,101 @@ export class ControlGroup extends AbstractControl {
...
+    });
+
+    super.markAsTouched();
+  }
+
+  /**
+   * Mark the group and all it's children as untouched.
+   */
+  markAllAsUntouched(): void {
+    StringMapWrapper.forEach(this.controls, (control, name) => {
```
Comment (vsavkin) [inline modules/angular2/src/common/forms/model.ts]: Same comment about the if statement.

File: modules/angular2/src/common/forms/model.ts [hunk 5]
```diff
@@ -367,6 +378,101 @@ export class ControlGroup extends AbstractControl {
...
+  }
+
+  /**
+   * Mark the group and all it's children as pending.
+   */
+  markAllAsPending({onlySelf}: {onlySelf?: boolean} = {}): void {
+    onlySelf = normalizeBool(onlySelf);
+
+    StringMapWrapper.forEach(this.controls, (control, name) => {
+      if (control instanceof Control) {
```
Comment (vsavkin) [inline modules/angular2/src/common/forms/model.ts]: Same comment about the if statement.

File: modules/angular2/src/common/forms/model.ts [hunk 6]
```diff
@@ -476,6 +582,108 @@ export class ControlArray extends AbstractControl {
...
+  }
+
+  /**
+   * Mark the array and all it's children as pristine.
+   */
+  markAllAsPristine({onlySelf}: {onlySelf?: boolean} = {}): void {
+    onlySelf = normalizeBool(onlySelf);
+
+    for (var i = 0; i < this.controls.length; i++) {
+      var control = this.controls[i];
```
Comment (vsavkin) [inline modules/angular2/src/common/forms/model.ts]: Same comment about the if statement.

File: modules/angular2/src/common/forms/model.ts [hunk 7]
```diff
@@ -476,6 +582,108 @@ export class ControlArray extends AbstractControl {
...
+
+  /**
+   * Mark the array and all it's children as dirty.
+   */
+  markAllAsDirty({onlySelf}: {onlySelf?: boolean} = {}): void {
+    onlySelf = normalizeBool(onlySelf);
+
+    for (var i = 0; i < this.controls.length; i++) {
+      var control = this.controls[i];
+      if (control instanceof Control) {
```
Comment (vsavkin) [inline modules/angular2/src/common/forms/model.ts]: Same comment about the if statement.

File: modules/angular2/src/common/forms/model.ts [hunk 8]
```diff
@@ -476,6 +582,108 @@ export class ControlArray extends AbstractControl {
...
+    super.markAsDirty({onlySelf: onlySelf});
+  }
+
+  /**
+   * Mark the array and all it's children as touched.
+   */
+  markAllAsTouched(): void {
+    for (var i = 0; i < this.controls.length; i++) {
+      var control = this.controls[i];
+      if (control instanceof Control) {
```
Comment (vsavkin) [inline modules/angular2/src/common/forms/model.ts]: Same comment about the if statement.

File: modules/angular2/src/common/forms/model.ts [hunk 9]
```diff
@@ -476,6 +582,108 @@ export class ControlArray extends AbstractControl {
...
+    }
+
+    super.markAsTouched();
+  }
+
+  /**
+   * Mark the array and all it's children as untouched.
+   */
+  markAllAsUntouched(): void {
+    for (var i = 0; i < this.controls.length; i++) {
```
Comment (vsavkin) [inline modules/angular2/src/common/forms/model.ts]: Same comment about the if statement.

File: modules/angular2/src/common/forms/model.ts [hunk 10]
```diff
@@ -476,6 +582,108 @@ export class ControlArray extends AbstractControl {
...
+  }
+
+  /**
+   * Mark the array and all it's children as pending.
+   */
+  markAllAsPending({onlySelf}: {onlySelf?: boolean} = {}): void {
+    onlySelf = normalizeBool(onlySelf);
+
+    for (var i = 0; i < this.controls.length; i++) {
+      var control = this.controls[i];
```
Comment (vsavkin) [inline modules/angular2/src/common/forms/model.ts]: Same comment about the if statement.

**Comment by basvandenheuvel** [comment]
@vsavkin I'm improving the PR right now, thanks for looking in to this PR.

I have a few questions for future PR's since i like to continue on improving the Forms part.

1. Methods like "dirty" and "pristine" bubble up to the parent, the touched part doesn't. Is this intended? If the touched part should also bubble up i can create a PR for this.

2. Right now the ControlArray and ControlGroup also extend from AbstractControl. Because of this the group and array also have the properties like touched, pristine etc etc. If you think about it, a group or array doesn't need this. All that a group or array should do if we asked it or it is pristine, is check if all it's children are pristine and return the correct boolean. 
Because of this i'm thinking about dividing the stuff within AbstractControl into two different abstract classes. Control will extend from both, and group and array only from one.
I wonder what you think about this idea.

**Comment by basvandenheuvel** [comment]
@vsavkin  Tests are currently failing: https://travis-ci.org/basvandenheuvel/angular/jobs/106937621#L1016 , i'm not sure why that is. I did add all methods to the common.ts in the past.

**Comment by vsavkin** [comment]
You need to update `angular/tools/public_api_guard/public_api_spec.ts`

**Comment by vsavkin** [comment]
>> Methods like "dirty" and "pristine" bubble up to the parent, the touched part doesn't. Is this intended? If the touched part should also bubble up i can create a PR for this.

It should bubble up.

>> Right now the ControlArray and ControlGroup also extend from AbstractControl. Because of this the group and array also have the properties like touched, pristine etc etc. If you think about it, a group or array doesn't need this. All that a group or array should do if we asked it or it is pristine, is check if all it's children are pristine and return the correct boolean. 

You are probably right. It does not change the public api, so let's not do it right now.

**Comment by basvandenheuvel** [comment]
@vsavkin  I "squashed" all my commits into one. It also took all the merge remote-tracking branch stuff. After i did that i had to update again and had a lot of conflicts i had to solve. Looks like i got everything right now.

Because i squashed all the merge remote-tracking branch stuff it says i changed a lot of files i didn't change myself.

**Comment by louisscruz** [comment]
+1

**Comment by vsavkin** [comment]
I am confused. Why is the second commit required?

**Comment by basvandenheuvel** [comment]
@vsavkin i'm even more confused. After the squash i did get A  LOT of merge errors. The second commit fixes them....

**Comment by basvandenheuvel** [comment]
@vsavkin it does look weird........ shall i just create a new PR and copy the changes i made?

**Comment by basvandenheuvel** [comment]
new pr: https://github.com/angular/angular/pull/7288

**Pull request closed without merging**